File: backend/src/hypomnema/llm/google.py
"""Google (Gemini) LLM client."""

from typing import Any

from google import genai
from google.genai import types

from hypomnema.llm.json_utils import parse_json_object


class GoogleLLMClient:
    DEFAULT_MODEL = "gemini-2.5-flash"

    def __init__(self, api_key: str, *, model: str = "") -> None:
        self._client = genai.Client(api_key=api_key)
        self._model = model or self.DEFAULT_MODEL

    @staticmethod
    def _build_config(
        *,
        system: str = "",
        json_mode: bool = False,
        timeout_ms: int | None = None,
    ) -> types.GenerateContentConfig | None:
        http_options = types.HttpOptions(
            retry_options=types.HttpRetryOptions(attempts=1),
        )
        if timeout_ms is not None:
            http_options.timeout = timeout_ms

        kwargs: dict[str, object] = {
            "http_options": http_options,
        }
        if system:
            kwargs["system_instruction"] = system
        if json_mode:
            kwargs["response_mime_type"] = "application/json"
        return types.GenerateContentConfig(**kwargs)

    async def complete(self, prompt: str, *, system: str = "", timeout_ms: int | None = None) -> str:
        config = self._build_config(system=system, timeout_ms=timeout_ms)
        response = await self._client.aio.models.generate_content(
            model=self._model,
            contents=prompt,
            config=config,
        )
        if response.text is None:
            raise ValueError("Empty response from Google API")
        return response.text

    async def complete_json(
        self,
        prompt: str,
        *,
        system: str = "",
        timeout_ms: int | None = None,
    ) -> dict[str, Any]:
        config = self._build_config(
            system=system,
            json_mode=True,
            timeout_ms=timeout_ms,
        )
        response = await self._client.aio.models.generate_content(
            model=self._model,
            contents=prompt,
            config=config,
        )
        if response.text is None:
            raise ValueError("Empty response from Google API")
        return parse_json_object(response.text)

    async def count_tokens(self, text: str) -> int:
        response = await self._client.aio.models.count_tokens(
            model=self._model,
            contents=text,
        )
        total_tokens = response.total_tokens
        if total_tokens is None:
            raise ValueError("Empty token count from Google API")
        return total_tokens

File: server/internal/llm/google.go
package llm

import (
	"context"
	"encoding/json"
	"fmt"
)

type google struct {
	model  string
	apiKey string
}

func newGoogle(model, apiKey string) *google {
	if model == "" {
		model = "gemini-2.5-flash"
	}
	return &google{model: model, apiKey: apiKey}
}

func (g *google) Provider() string { return "google" }

func (g *google) Complete(ctx context.Context, prompt, system string) (string, error) {
	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/%s:generateContent?key=%s",
		g.model, g.apiKey)

	contents := []map[string]any{
		{"role": "user", "parts": []map[string]string{{"text": prompt}}},
	}
	body := map[string]any{"contents": contents}
	if system != "" {
		body["systemInstruction"] = map[string]any{
			"parts": []map[string]string{{"text": system}},
		}
	}

	respData, err := httpDo(ctx, "POST", url, nil, body)
	if err != nil {
		return "", fmt.Errorf("google: %w", err)
	}

	var resp struct {
		Candidates []struct {
			Content struct {
				Parts []struct {
					Text string `json:"text"`
				} `json:"parts"`
			} `json:"content"`
		} `json:"candidates"`
	}
	if err := json.Unmarshal(respData, &resp); err != nil {
		return "", fmt.Errorf("google parse: %w", err)
	}
	if len(resp.Candidates) == 0 || len(resp.Candidates[0].Content.Parts) == 0 {
		return "", fmt.Errorf("google: empty response")
	}
	return resp.Candidates[0].Content.Parts[0].Text, nil
}

func (g *google) CompleteJSON(ctx context.Context, prompt, system string) (map[string]any, error) {
	text, err := g.Complete(ctx, prompt, system)
	if err != nil {
		return nil, err
	}
	return extractJSON(text)
}
